### PR TITLE
Consolidate Custom Filter Logic

### PR DESCRIPTION
**Pull request by fa9r**
## Describe changes
I consolidated how custom filters (that are not based on the columns of the underlying table) are applied to filter models by introducing a `BaseFilter.get_custom_filters()` method that all subclasses with custom filter logic now override (instead of defining the custom filters inside `generate_filter()` or in the respective `SQLZenStore` methods).

As part of this I also adjusted all string custom filter columns to support advanced filtering operators, e.g. `list_model_version_artifact_links(artifact_name="contains:abc")`

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/stacks-and-components/component-guide) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.
- [x] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: src/zenml/models/v2/core/model_version_artifact.py [hunk 1]
```diff
@@ -164,15 +179,54 @@ class ModelVersionArtifactFilter(WorkspaceScopedFilter):
...
+            artifact_name_filter = and_(  # type: ignore[type-var]
+                ModelVersionArtifactSchema.artifact_version_id
+                == ArtifactSchema.id,
+                filter_.generate_query_conditions(ArtifactSchema),
+            )
+            custom_filters.append(artifact_name_filter)
+
+        if self.only_data_artifacts:
+            data_artifact_filter = and_(  # type: ignore[type-var]
+                ModelVersionArtifactSchema.is_model_artifact is False,
```
[the remark this answers is not in the document]
Comment (fa9r) [reply src/zenml/models/v2/core/model_version_artifact.py:215]: Yes good catch!

**Comment by fa9r** [comment]
@schustmi I manually tested all affected filter models, should be good now. 

While at it, I also introduced some merge conflicts for your #2103 PR :)

**Comment by schustmi** [comment]
Was just about to mention that 😄 
Mine will be merged anytime now, so we'll have to adjust on this PR

File: src/zenml/models/v2/core/artifact_version.py [hunk 2]
```diff
@@ -390,3 +392,51 @@ class ArtifactVersionFilter(WorkspaceScopedFilter):
...
+            ArtifactVersionSchema,
+        )
+        from zenml.zen_stores.schemas.step_run_schemas import (
+            StepRunInputArtifactSchema,
+            StepRunOutputArtifactSchema,
+        )
+
+        if self.name:
+            value, filter_operator = self._resolve_operator(self.name)
+            filter_ = StrFilter(
```
[the remark this answers is not in the document]
Comment (fa9r) [reply src/zenml/models/v2/core/artifact_version.py:429]: Yep exactly :)

Regarding sorting, we already have that implemented on `filter.py` L336:

```python
if column in cls.FILTER_EXCLUDE_FIELDS:
    raise ValueError(
        f"This resource can not be sorted by this field: '{v}'"
    )
```

**Comment by fa9r** [comment]
CI template test failures can be ignored since they passed here: https://github.com/zenml-io/template-e2e-batch/actions/runs/7125493871?pr=27

**Review by schustmi — APPROVED** [review]
LGTM

**Review by bcdurak — APPROVED** [review]
🦭

**Pull request merged**